docs/standard/serialization/system-text-json-migrate-from-newtonsoft-how-to.md

```diff
@@ -0,0 +1,553 @@
+
+Since the `Utf8JsonReader` considers the input to be JSON text, a UTF-8 byte order mark (BOM) is considered invalid JSON. The caller needs to filter that out before passing the data to the reader.
+
+For code examples, see [Use Utf8JsonReader](system-text-json-how-to.md#use-utf8jsonreader).
+
+### Read with multi-segment ReadOnlySequence
+
+If your JSON input is a [ReadOnlySpan\<byte>](xref:System.ReadOnlySpan%601), each JSON element can be accessed from the `ValueSpan` property on the reader as you go through the read loop. However, if your input is a [ReadOnlySequence\<byte>](xref:System.Buffers.ReadOnlySequence%601) (which is the result of reading from a <xref:System.IO.Pipelines.PipeReader>), some JSON elements might straddle multiple segments of the `ReadOnlySequence<byte>` object. These elements would not be accessible from <xref:System.Text.Json.Utf8JsonReader.ValueSpan%2A> in a contiguous memory block. Instead, whenever you have a multi-segment `ReadOnlySequence<byte>` as input, poll the <xref:System.Text.Json.Utf8JsonReader.HasValueSequence%2A> property on the reader to figure out how to access the current JSON element. Here's a recommended pattern:
+
+```csharp
+while (reader.Read())
+{
+    switch (reader.TokenType)
+    {
+        // ...
+        ReadOnlySpan<byte> jsonElement = reader.HasValueSequence ?
+            reader.ValueSequence.ToArray() :
+            reader.ValueSpan;
+        // ...
+    }
+}
+```
+
+### Use ValueTextEquals for property name lookups
+
+Don't use <xref:System.Text.Json.Utf8JsonReader.ValueSpan%2A> to do byte-by-byte comparisons by calling <xref:System.MemoryExtensions.SequenceEqual%2A> for property name lookups. Call <xref:System.Text.Json.Utf8JsonReader.ValueTextEquals%2A> instead, because that method unescapes any characters that are escaped  in the JSON. Here's an example:
+
+[!code-csharp[](~/samples/snippets/core/system-text-json/csharp/Utf8ReaderFromFile.cs?highlight=10,37)]
+
+### Read null values into nullable value types
+
+`Newtonsoft.Json` provides APIs like `ReadAsBoolean`, which handle a `Null` `TokenType` for you by returning a `bool?`. The built-in `System.Text.Json` APIs return only non-nullable value types. For example, <xref:System.Text.Json.Utf8JsonReader.GetBoolean?displayProperty=nameWithType> returns a `bool`. It throws an exception if it finds `Null` in the JSON. The following examples show two ways to handle nulls, one by returning a nullable value type and one by returning the default value:
+
+```csharp
+public bool? ReadAsNullableBoolean()
+{
+    _reader.Read();
+    if (_reader.TokenType == JsonTokenType.Null)
+    {
+        return null;
+    }
+    if (_reader.TokenType != JsonTokenType.True && _reader.TokenType != JsonTokenType.False)
+    {
+        throw CreateUnexpectedException(ref _reader, "a JSON true or false literal token");
```

Comment:
Interesting. I might have missed it but what does `CreateUnexpectedException` do? Where is the code for this?